### PR TITLE
cephadm: ingress service for RGW

### PR DESCRIPTION
**Pull request by liewegas**
This is a rewrite of the HA_RGW service that manages a combination of haproxy + keepalived.

- it now works (HA_RGW was broken by the recent scheduler changes)
- much easier to configure (fewer options; more opinionated setup)
- scheduler changes to handle the "sidecar" placement of keepalived containers

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/schedule.py [hunk 1]
```diff
@@ -123,6 +124,35 @@ def validate(self) -> None:
                     f'Cannot place {self.spec.one_line_str()}: No matching '
                     f'hosts for label {self.spec.placement.label}')
 
+    def place_sidecars(
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py:127]: in my view, sidecars are containers that are places within systemd units. If I see this correctly, we're talking about placing daemons next to other daemons, right?

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -1053,12 +1055,20 @@ def __init__(self,
...
     def service_type(self) -> str:
         return self.spec.service_type
 
     def __repr__(self) -> str:
         return f"<ServiceDescription of {self.spec.one_line_str()}>"
 
+    def get_port_summary(self) -> str:
+        if not self.ports:
+            return ''
+        return f"{self.virtual_ip or '?'}:{','.join(map(str, self.ports or []))}"
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:1070]: in case of no virtual ip, should we add the `host addr` here? cc @varshar16

**Review by sebastian-philipp — COMMENTED** [review]
looks sane.

File: src/pybind/mgr/cephadm/inventory.py [hunk 3]
```diff
@@ -263,7 +263,7 @@ def __init__(self, mgr):
         self.last_facts_update = {}    # type: Dict[str, datetime.datetime]
         self.osdspec_previews = {}     # type: Dict[str, List[Dict[str, Any]]]
         self.osdspec_last_applied = {}  # type: Dict[str, Dict[str, datetime.datetime]]
-        self.networks = {}             # type: Dict[str, Dict[str, List[str]]]
+        self.networks = {}             # type: Dict[str, Dict[str, Dict[str, List[str]]]]
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/pybind/mgr/cephadm/inventory.py:266]: it's part of a larger json blob that's rebuilt on each save; will clean itself up

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (liewegas) [reply src/pybind/mgr/orchestrator/_interface.py:1070]: At this level, I think no.  ``orch ps`` output might include an IP for a specific daemon instance, but at the service level, we can/should only show an IP here if there is a stable (virtual) IP for the service that is independent of daemon (re)placement, scheduling, etc.

File: src/pybind/mgr/cephadm/schedule.py (hunk 1 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/schedule.py]: renamed

**Comment by sebastian-philipp** [comment]
can you add a note to https://docs.ceph.com/en/latest/cephadm/compatibility/#stability that we're introducing breaking changes for unstable services?

**Comment by liewegas** [comment]
https://pulpito.ceph.com/sage-2021-04-13_17:36:49-rados:cephadm-wip-sage-testing-2021-04-13-1100-distro-basic-smithi/

**Review by sebastian-philipp — APPROVED** [review]

**Pull request merged**